tests/network/services.go

```diff
@@ -0,0 +1,321 @@
+	}
+
+	cleanupVMI := func(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance) {
+		if vmi != nil {
+			By("Deleting the VMI")
+			Expect(virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Delete(vmi.GetName(), &k8smetav1.DeleteOptions{})).To(Succeed())
+
+			By("Waiting for the VMI to be gone")
+			Eventually(func() error {
+				_, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmi.GetName(), &k8smetav1.GetOptions{})
+				return err
+			}, 2*time.Minute, time.Second).Should(SatisfyAll(HaveOccurred(), WithTransform(errors.IsNotFound, BeTrue())), "The VMI should be gone within the given timeout")
+		}
+	}
+
+	assertConnectionToExposedServiceSucceeds := func(serviceName, namespace string, servicePort int) {
+		serviceFQDN := fmt.Sprintf("%s.%s", serviceName, namespace)
+
+		By(fmt.Sprintf("starting a job which tries to reach the vmi via service %s", serviceFQDN))
+		job := runTCPClientExpectingHelloWorldFromServer(serviceFQDN, strconv.Itoa(servicePort), namespace)
+
+		By(fmt.Sprintf("waiting for the job to report a SUCCESSFUL connection attempt to service %s on port %d", serviceFQDN, servicePort))
+		tests.WaitForJobToSucceed(job, 90)
+	}
+
+	assertConnectionToExposedServiceFails := func(serviceName, namespace string, servicePort int) {
+		serviceFQDN := fmt.Sprintf("%s.%s", serviceName, namespace)
+
+		By(fmt.Sprintf("starting a job which tries to reach the vmi via service %s", serviceFQDN))
+		job := runTCPClientExpectingHelloWorldFromServer(serviceFQDN, strconv.Itoa(servicePort), namespace)
+
+		By(fmt.Sprintf("waiting for the job to report a FAILED connection attempt to service %s on port %d", serviceFQDN, servicePort))
+		tests.WaitForJobToFail(job, 90)
+	}
+
+	BeforeEach(func() {
+		var err error
+		virtClient, err = kubecli.GetKubevirtClient()
+		Expect(err).NotTo(HaveOccurred(), "Should successfully initialize an API client")
+	})
+
+	Context("bridge interface binding", func() {
+		var inboundVMI *v1.VirtualMachineInstance
+		var serviceName string
+
+		const (
+			selectorLabelKey   = "expose"
+			selectorLabelValue = "me"
+			servicePort        = 1500
+		)
+
+		createVMISpecWithBridgeInterface := func() *v1.VirtualMachineInstance {
+			return libvmi.NewCirros(
+				libvmi.WithInterface(libvmi.InterfaceDeviceWithBridgeBinding()),
+				libvmi.WithNetwork(v1.DefaultPodNetwork()))
+		}
+
+		createReadyVMIWithBridgeBindingAndExposedService := func(hostname string, subdomain string) *v1.VirtualMachineInstance {
+			return readyVMI(
+				exposeExistingVMISpec(
+					createVMISpecWithBridgeInterface(), subdomain, hostname, selectorLabelKey, selectorLabelValue))
+		}
+
+		BeforeEach(func() {
+			subdomain := "vmi"
+			hostname := "inbound"
+
+			inboundVMI = createReadyVMIWithBridgeBindingAndExposedService(hostname, subdomain)
+			tests.StartTCPServer(inboundVMI, servicePort)
+		})
+
+		AfterEach(func() {
+			cleanupVMI(virtClient, inboundVMI)
```

Comment:
Assigning `inboundVMI` variable is blindly done in `BeforeEach` block. 

All `AfterEach` needs to do is make sure the entity is gone (if it existed).

Reply:
> lets say NewCirros which is part of creating a new inboundVMI uses new which has an exception,
> in this situation, in case inboundVMI isnt re-declared on each iteration of `Context`
> it will still have the previous value it had from the previous run.

I might be missing something here; what is the issue here ? AFAIU, what you want is the VMI gone on the test tear-down. In the scenario you're presenting, the VMI was never even created - i.e. there's nothing to clean.

> 
> In any case, i dont think we should assume a blind success, even if its the "first" thing that is done.
> Code might be changed, Exceptions can be added, and its already should be right at this current state imo.

Reply:
The inboundVMI var is not redeclared each test by the `var` line which initialize it - correct me if i am wrong in this because this is the assumption for the rest of the test.

Now in case you have 2 tests, and one is created + deleted the VMI, but didnt set inboundVMI to nil,
the next test, in case of whatever failure / assert, before the VMI was created or the inboundVMI set to be nil (for example if new failed and returned nil but didnt assert), would retry to delete the already deleted VMI.

in case this can happen, only needed line is to add inboundVMI = nil in the end of cleanupVMI.

Reply:
> The inboundVMI var is not redeclared each test by the `var` line which initialize it - correct me if i am wrong in this because this is the assumption for the rest of the test.
> 
> Now in case you have 2 tests, and one is created + deleted the VMI, but didnt set inboundVMI to nil,
> the next test, in case of whatever failure / assert, before the VMI was created or the inboundVMI set to be nil (for example if new failed and returned nil but didnt assert), would retry to delete the already deleted VMI.

Which would throw an error in the tear-down, which doesn't matter, since by now the test has already failed, and the VM does not exist. Meaning that the end result is the same: test failed, no VM present. But with less code.

> 
> in case this can happen, only needed line is to add inboundVMI = nil in the end of cleanupVMI.

Reply:
imo its cleaner and better if it would know to handle it and not create errors that can distract the debugger

Reply:
```
			var selectedNode *k8sv1.Node

			BeforeEach(func() {
				selectedNode = nil
			})
```

You can even see other tests are doing something like this, that i would suggest please to adapt, just adding the nil thats all needed.

Reply:
> ```
> 			var selectedNode *k8sv1.Node
> 
> 			BeforeEach(func() {
> 				selectedNode = nil
> 			})
> ```
> 

I don't have a fair chance of knowing where that snippet came from; for all I know, you could have written that yourself to validate your narrative (I'm **not** saying you did that). For future times, please just paste a link - so I can get context of what you're trying to show. Which, is impossible to infer from the snippet above.

Nevertheless, I can also show parts where that is not done (and in the network test package):
  - https://github.com/kubevirt/kubevirt/blob/master/tests/network/primary_pod_network.go#L113

I just followed the current status-quo of the network test package.

Since I'm looting the cleanup function from the `primary_pod_network` test, I want to propose a follow up PR unifying the VM creation (and waiting for readiness) and cleanup stages of it. I welcome you to move your complaint there, in due time.

> You can even see other tests are doing something like this, that i would suggest please to adapt, just adding the nil thats all needed.

I think I've explained in https://github.com/kubevirt/kubevirt/pull/4035#discussion_r484727019 why resetting the vmi is **not** needed. It achieves the exact same thing, on all scenarios. You didn't refute my claims, just said "imo its cleaner and better if it would know to handle it".

I prefer to do the exact minimum for it to work.

Reply:
i didnt agree with you, i dont want it to throw an error that can be eliminated
and it took it from "tests/infra_test.go"
if i said that its from the code its from the code.
https://github.com/kubevirt/kubevirt/blob/bf5b5e5d50073b2f7ede036531609280343e6664/tests/infra_test.go#L220

Reply:
> i didnt agree with you, i dont want it to throw an error that can be eliminated

The error it will throw is the error **where the VM failed to be created**, not the one deleting a VMI that doesn't exist. I've said that before.

Since it seems I'm not getting through to you with words, I'll try to show an example. Just execute the following snippet:

```golang
var _ = Describe("errors on after each do NOT bother anyone when stuff explodes on the test setup ", func() {
	BeforeEach(func() {
		Expect(2).NotTo(Equal(2), "ERROR in the SETUP")
	})

	It("will NOT even be executed", func(){
		Expect(2).To(Equal(2), "not executed ...")
	})

	AfterEach(func() {
		Expect(2).NotTo(Equal(2), "ERROR in the TEARDOWN")
	})
})
```

If you run this example, you'll see the following output: 

```
• Failure in Spec Setup (BeforeEach) [0.001 seconds]
errors on after each do NOT bother anyone when stuff explodes on the test setup 
/root/go/src/kubevirt.io/kubevirt/tests/probes_test.go:19
  will NOT even be executed [BeforeEach]
  /root/go/src/kubevirt.io/kubevirt/tests/probes_test.go:24

  ERROR in the SETUP
  Expected
      <int>: 2
  not to equal
      <int>: 2

  /root/go/src/kubevirt.io/kubevirt/tests/probes_test.go:21

```

Since it does not throw the error in the after each, what is the point of setting it to null ? It's a redundant operation.

> and i took it from "tests/infra_test.go"
> if i said that its from the code, and pasted a snippet, then its from the code.

Well, I was told to not trust anyone. Which is good advice. You're not an exception. 

We can at least agree that if you paste whatever random snippet, whoever reads it has no context about it, right ? 

Context is important (for instance, I would say you're right - **if** this code was **outside** a beforeEach / afterEach blocks). 

> Anyway sure here is the link: (just thought that the snippet i pasted was a self contained real life example)
> https://github.com/kubevirt/kubevirt/blob/bf5b5e5d50073b2f7ede036531609280343e6664/tests/infra_test.go#L220
> 
> About
> https://github.com/kubevirt/kubevirt/blob/master/tests/network/primary_pod_network.go#L113
> I think i even myself did the same mistake on a 3rd file, and still think its a good practice and coding that we will fix them.

As I've said; I don't think you/whoever wrote that code did any mistake.

Reply:
> > i didnt agree with you, i dont want it to throw an error that can be eliminated
> 
> The error it will throw is the error **where the VM failed to be created**, not the one deleting a VMI that doesn't exist. I've said that before.
> 
> Since it seems I'm not getting through to you with words, I'll try to show an example. Just execute the following snippet:
> 
> ```go
> var _ = Describe("errors on after each do NOT bother anyone when stuff explodes on the test setup ", func() {
> 	BeforeEach(func() {
> 		Expect(2).NotTo(Equal(2), "ERROR in the SETUP")
> 	})
> 
> 	It("will NOT even be executed", func(){
> 		Expect(2).To(Equal(2), "not executed ...")
> 	})
> 
> 	AfterEach(func() {
> 		Expect(2).NotTo(Equal(2), "ERROR in the TEARDOWN")
> 	})
> })
> ```
> 
> If you run this example, you'll see the following output:
> 
> ```
> • Failure in Spec Setup (BeforeEach) [0.001 seconds]
> errors on after each do NOT bother anyone when stuff explodes on the test setup 
> /root/go/src/kubevirt.io/kubevirt/tests/probes_test.go:19
>   will NOT even be executed [BeforeEach]
>   /root/go/src/kubevirt.io/kubevirt/tests/probes_test.go:24
> 
>   ERROR in the SETUP
>   Expected
>       <int>: 2
>   not to equal
>       <int>: 2
> 
>   /root/go/src/kubevirt.io/kubevirt/tests/probes_test.go:21
> ```
> 
> Since it does not throw the error in the after each, what is the point of setting it to null ? It's a redundant operation.
> 
> > and i took it from "tests/infra_test.go"
> > if i said that its from the code, and pasted a snippet, then its from the code.
> 
> Well, I was told to not trust anyone. Which is good advice. You're not an exception.
> 
> We can at least agree that if you paste whatever random snippet, whoever reads it has no context about it, right ?
> 
> Context is important (for instance, I would say you're right - **if** this code was **outside** a beforeEach / afterEach blocks).

Thanks for the detailed explanation.
Lets say i agreed for everything else for the sake of this:
So cleanupVMI is limited to be used only on beforeEach afterEach blocks, and is not re-entrant in the matter of the pointer atm, which is wrong imo. (instead just adding in the function itself a set to nil in case the the clean was executed as i told in the very beginning)
This is my opinion at least, i wont block anything.

> 
> > Anyway sure here is the link: (just thought that the snippet i pasted was a self contained real life example)
> > https://github.com/kubevirt/kubevirt/blob/bf5b5e5d50073b2f7ede036531609280343e6664/tests/infra_test.go#L220
> > 
> > About
> > https://github.com/kubevirt/kubevirt/blob/master/tests/network/primary_pod_network.go#L113
> > I think i even myself did the same mistake on a 3rd file, and still think its a good practice and coding that we will fix them.
> 
> As I've said; I don't think you/whoever wrote that code did any mistake.

Reply:
> I find this an interesting discussion.
> To summarize, if a variable is defined at a container body, it is initialized at the definition, but this definition only occurs once for all the `It` tests under it.
> If we have one test, there is no problem.
> If we have two or more test, the variable will not be re-initialize per test unless explicitly set.
> 
> Therefore, if at the first test, an assert fails post this variable being populated, and the next test fails before the variable is set, the second test teardown may process an older value.
> 
> I agree that this is sometime a non-issue, but at the same time it may become a real issue without anyone noticing.
> But there is an even more interesting outcome here: If in some scenarios, the variable check against a nil is not assuring that we already created the resource, and we are fine with attempting to delete it anyway, then there is no reason to check against the nil in the first place.

I can remove the check, thus making the caller responsible for assuring the VMI exists. IMHO, this would indeed be the best.

> 
> Continuing this, in case we do re-initialize the variable, where this needs to occur?
> If we do this at the teardown, we need to make sure not to nil it too early. If the cleanup operation fails, should we nil it or not?
> It could also be done at the setup,e.g. re-initialize each time.
> 
> Bottom line, this ginkgo framework is presenting a real problem, and we surely will not solve it in this PR or this discussion.

Reply:
Miguel, Edy
What about this?
```
				if inboundVMI != nil {
					defer func() {
						inboundVMI = nil
					}()
```